plan/task/clientfilesystem.go

```diff
@@ -0,0 +1,61 @@
+package task
+
+import (
+	"context"
+	"crypto/sha256"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/gofrs/flock"
+	"github.com/mitchellh/go-homedir"
+	"github.com/rs/zerolog/log"
+	"go.dagger.io/dagger/compiler"
+	"go.dagger.io/dagger/plancontext"
+)
+
+const clientFsTempKey = "client-filesystem-lock"
+
+// clientFsPath extracts the absolute file path from a client filesystem path value.
+func clientFsPath(v *compiler.Value) (string, error) {
+	path, err := v.String()
+	if err != nil {
+		return "", err
+	}
+	expanded, err := homedir.Expand(path)
+	if err != nil {
+		return "", err
+	}
+	return filepath.Abs(expanded)
+}
+
+// clientFsLock acquires a file lock to read/write a client filesystem resource.
+// Every path gets it's own lock. This allows multiple read/write to the same
+// resource, protecting against race conditions.
+func clientFsLock(ctx context.Context, pctx *plancontext.Context, fsPath string) (*flock.Flock, error) {
+	lg := log.Ctx(ctx)
+
+	// Use a temporary dir for this session so it gets cleaned up in the end
+	dir, ok := pctx.TempDirs.Get(clientFsTempKey)
+	if !ok {
```

Comment:
race condition here if we access multiple FS at the same time (which we are). We may end up losing lock files

Reply:
That's right, good catch 😅